src/IR.jl

```diff
@@ -0,0 +1,1127 @@
+    if mlirIsNull(mlir_dialect)
+        error("could not load dialect from handle $handle")
+    else
+        Dialect(mlir_dialect)
+    end
+end
+function get_or_load_dialect!(context, dialect::String)
+    get_or_load_dialect!(context, DialectHandle(Symbol(dialect)))
+end
+
+function enable_multithreading!(context, enable=true)
+    API.mlirContextEnableMultithreading(context, enable)
+    context
+end
+
+is_registered_operation(context, opname) = API.mlirContextIsRegisteredOperation(context, opname)
+
+### Location
+
+struct Location
+    location::MlirLocation
+
+    Location(location) = begin
+        @assert !mlirIsNull(location) "cannot create Location with null MlirLocation"
+        new(location)
+    end
+end
+
+Location(context::Context) = Location(API.mlirLocationUnknownGet(context))
+Location(context::Context, filename, line, column) =
+    Location(API.mlirLocationFileLineColGet(context, filename, line, column))
+
+Base.convert(::Type{MlirLocation}, location::Location) = location.location
+
+function Base.show(io::IO, location::Location)
+    c_print_callback = @cfunction(print_callback, Cvoid, (MlirStringRef, Any))
+    ref = Ref(io)
+    print(io, "Location(#= ")
+    API.mlirLocationPrint(location, c_print_callback, ref)
+    print(io, " =#)")
+end
+
+### Type
+
+struct MLIRType
+    type::MlirType
+
+    MLIRType(type) = begin
+        @assert !mlirIsNull(type)
+        new(type)
+    end
+end
+
+MLIRType(t::MLIRType) = t
+MLIRType(context::Context, T::Type{<:Signed}) =
+    MLIRType(API.mlirIntegerTypeGet(context, sizeof(T) * 8))
+MLIRType(context::Context, T::Type{<:Unsigned}) =
+    MLIRType(API.mlirIntegerTypeGet(context, sizeof(T) * 8))
+MLIRType(context::Context, ::Type{Bool}) =
+    MLIRType(API.mlirIntegerTypeGet(context, 1))
+MLIRType(context::Context, ::Type{Float32}) =
+    MLIRType(API.mlirF32TypeGet(context))
+MLIRType(context::Context, ::Type{Float64}) =
+    MLIRType(API.mlirF64TypeGet(context))
+MLIRType(context::Context, ft::Pair) =
+    MLIRType(API.mlirFunctionTypeGet(context,
+        length(ft.first), [MLIRType(t) for t in ft.first],
+        length(ft.second), [MLIRType(t) for t in ft.second]))
+MLIRType(context, a::AbstractArray{T}) where {T} = MLIRType(context, MLIRType(context, T), size(a))
+MLIRType(context, ::Type{<:AbstractArray{T,N}}, dims) where {T,N} =
+    MLIRType(API.mlirRankedTensorTypeGetChecked(
+        Location(context),
+        N, collect(dims),
+        MLIRType(context, T),
+        Attribute(),
+    ))
+MLIRType(context, element_type::MLIRType, dims) =
+    MLIRType(API.mlirRankedTensorTypeGetChecked(
+        Location(context),
+        length(dims), collect(dims),
+        element_type,
+        Attribute(),
+    ))
+MLIRType(context, ::T) where {T<:Real} = MLIRType(context, T)
+MLIRType(_, type::MLIRType) = type
+
+IndexType(context) = MLIRType(API.mlirIndexTypeGet(context))
+
+Base.convert(::Type{MlirType}, mtype::MLIRType) = mtype.type
+
```

Comment:
Type parsing could be added analogous to `parse(::Type{MLIRAttribute})`
```suggestion
Base.parse(::Type{MLIRType}, context, s) =
    MLIRType(API.mlirTypeParseGet(context, s))
    
```